app/controllers/authenticate_controller.rb

```diff
@@ -1,30 +1,89 @@
 class AuthenticateController < ApplicationController
   include TokenGenerator
 
-  # Request path indicates the username for +authenticate+.
-  before_filter :credentials_lookup
+  class MemoizedRole
+    def self.[](role_id)
+      @user_roles ||= Hash.new { |h, id| h[id] = Role[id] }
+      @user_roles[role_id]
+    end
+
+    def self.roleid_from_username(account, username)
+      Role.roleid_from_username(account, username)
+    end
+  end
 
   def authenticate
-    api_key = request.body.read
-    if @credentials.valid_api_key? api_key
-      render json: authentication_token
+
+    authenticator  = params[:authenticator]
+    service_id     = params[:service_id]
+    account        = params[:account]
+    username       = params[:id]
+    password       = request.body.read
+    role_id        = MemoizedRole.roleid_from_username(account, username)
+
+    validate_role_exists!(role_id)
+    
+    case authenticator
+    when 'authn-ldap'
+      validate_security!(authenticator, account, service_id, username)
+      raise Unauthorized unless ldap_authenticator.valid?(username, password)
+      role = MemoizedRole[role_id]
+    # default conjur: authn
     else
-      head :unauthorized
+      credentials = Credentials[role_id]
+      validate_credentials!(credentials, password)
+      role = credentials.role
     end
+
+    # If we arrive here, the user is authenticated and `role` is set
+    # Otherwise Unauthorized would have already been raised
+
+    authentication_token = sign_token(role)
+    render json: authentication_token
   end
   
   protected
-  
-  def authentication_token
-    sign_token @credentials.role
+
+  def validate_security!(authenticator, account, service_id, user_id)
+    security = AuthenticatorSecurity.new(authn_type: authenticator,
+      account: account, role_class: MemoizedRole)
+    security.validate(service_id, user_id)
+  rescue AuthenticatorNotEnabled, ServiceNotDefined, NotAuthorizedInConjur => e
+    logger.debug(e.message)
+    raise Unauthorized
+  rescue => e
+    logger.debug("Unexpected AuthenticatorSecurity Error: #{e.message}")
+    raise Unauthorized
   end
 
-  def credentials_lookup
-    roleid = Role.roleid_from_username(account, params[:id])
-    @credentials = Credentials[roleid]
-    unless @credentials
-      logger.debug "Credentials for #{roleid} not found"
+  def validate_role_exists!(role_id)
```

Comment:
this is just checking that the `role_id` isn't falsey, but to me `validate_role_exists` implies that we are checking in the DB that the role referenced by the role_id exists